### PR TITLE
#167730752 User Roles Setup

### PR DESCRIPTION
**Pull request by OlawaleJoseph**
#### What does this PR do?
  This PR Implements Role-based access control.

#### Description of Task to be completed?
  - Implement the functionality to assign a role to a user.
  - Ensure only logged in Super Administrators can assign roles to user.
  - Ensure a Super Admin cannot change his/her role.

#### How should this be manually tested?
  - git clone this [branch](https://github.com/andela/firestar-backend/tree/ft-user-roles-setup-167730752)
  - cd into the project
  - checkout to the branch using the command `git checkout ft-user-roles-setup-167730752`.
  - Create a Postgres Database
  - Create a `.env` file and set PORT, JWT_SECRET and DB variables as specified in the .env.example file.
  - Run the command `npm install` in the terminal.
  - Open a terminal and run `npm test`
  - After the tests have passed. Run `npm run start:dev` in the console.

    1. Open Postman, make a post request to `localhost:9003/api/v1/auth/login` with the 
         request object 
            {
               email: abc123@gmail.com
             }.
    2. Copy the token in the response
    3. Make a patch request to `localhost:9003/api/v1/users/user/role` with
          the token from above in the header as ` authorization`.
    4. Send a request object containing the following:
           roleId: A digit, currently between 1 to 5.
	   email: abc123@gmail.com
         An unauthorized error will be displayed.
    5. Repeat steps 1 - 4 with the request object for step 1 as:
         {
            email: barefoot@gmail.com
         }
         The user object with the updated role will be displayed.

#### Any background context you want to provide?
Presently the login functionality isn't provided in this project. A dummy login functionality was implemented to generate JWT. 

#### What are the relevant pivotal tracker stories?
[#167730752](https://www.pivotaltracker.com/story/show/167730752)

#### Screenshots (if appropriate)
Screenshot (42)

Screenshot (24)
#### Questions:
N/A

### CODE REVIEW

**Review by conquext — COMMENTED** [review]
Could you include a screenshot that captures the step 4 of setting permission?

**Comment by OlawaleJoseph** [comment]
> Could you include a screenshot that captures the step 4 of setting permission?

Done

File: src/__tests__/permisions.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+
+chai.use(chaiHttp);
+
+const { assert } = chai;
+
+describe('Set Role Permissions', () => {
+  let unauthorised, invalidRoleId, invalidResourceId, validInput, validInput2;
+  beforeEach(() => {
+    unauthorised = {
```
Comment (dewaleolaoye) [inline src/__tests__/permisions.js]: Nice Work!
I suggest you have the `unauthorised, invalidRoleId, invalidResourceId, validInput, validInput2` object in a separate file and import them here  for separation of concern

Comment (OlawaleJoseph) [reply src/__tests__/permisions.js]: Alright

Comment (OlawaleJoseph) [reply src/__tests__/permisions.js]: Done

**Comment by conquext** [comment]
Please fix this failing test
image

File: .env.example [hunk 2]
```diff
@@ -1,7 +1,7 @@
...
-PORT=
-
-DB_TEST_NAME=barefoot_test
+DB_USERNAME=database username
+DB_PASSWORD=database password
+DB_NAME=name of database for development
+DB_HOST=Host url
+PORT=port number to listen for connections
+DB_TEST_NAME=name of test database
+SECRET=secret variable to sign and decode tokens
```
Comment (zanio) [inline .env.example]: Be more descriptive about the name `SECRET`  There are many secret key pair values in others people branch. Something like `SECRET_USER_ROLES_TOKEN`

File: package.json [hunk 3]
```diff
@@ -4,8 +4,8 @@
   "description": "A Social platform for the creative at heart",
   "main": "index.js",
   "scripts": {
-    "test": "cross-env NODE_ENV=development nyc mocha --require @babel/register ./src/__tests__/*.js",
```
Comment (zanio) [inline package.json]: WHY Did you change NODE_ENV to test

File: src/__tests__/index.js [hunk 4]
```diff
@@ -1,10 +0,0 @@
-import { equal } from 'assert';
-import arrayTest from '../utils/index';
-
-describe('Array', () => {
```
Comment (zanio) [inline src/__tests__/index.js]: Why was this file deleted ?

**Review by zanio — CHANGES_REQUESTED** [review]

File: .env.example (hunk 2 above)
Comment (OlawaleJoseph) [reply .env.example]: Thanks for the comment. That was just to get tokens. Creating of token is decided from the sign in route. That variable is a temporary one, pending when sign in route will be implemented.

**Comment by OlawaleJoseph** [comment]
> Please fix this failing test
> image

@conquext Thanks for the review. Kindly run npm run undo:test before running the test.
The reason being that, there is a function that updates or create the permissions of a role depending if the role already has the permissions.

**Comment by conquext** [comment]
After i run npm undo:test, 8 tests now failed

image

I think you should let your code manages the db state it requires by itself. you can create a beforeall hook with mocha to handle this

**Review by zanio — COMMENTED** [review]
image

Why Am i getting a test a failure after following the instruction on how it can be locally tested ? Is there something you are not providing in your `PR`?

**Comment by conquext** [comment]
> 
> Why Am i getting a test a failure after following the instruction on how it can be locally tested ? Is there something you are not providing in your `PR`?

Is it all of the tests that is failing? @zanio 
I think the port the test is run is listening to another service.

File: package.json (hunk 3 above)
Comment (OlawaleJoseph) [reply package.json]: NODE_ENV should be set to test when testing.

**Comment by OlawaleJoseph** [comment]
> After i run npm undo:test, 8 tests now failed
> 
> image
> 
> I think you should let your code manages the db state it requires by itself. you can create a beforeall hook with mocha to handle this

I Have fixed this

**Comment by OlawaleJoseph** [comment]
> image
> 
> Why Am i getting a test a failure after following the instruction on how it can be locally tested ? Is there something you are not providing in your `PR`?

The PORT you are trying to listen on is already connected to something else. Kidly close all previous connections or change the PORT number

File: src/__tests__/index.js (hunk 4 above)
Comment (zanio) [reply src/__tests__/index.js]: @OlawaleJoseph  you haven't reponded to this yet

Comment (OlawaleJoseph) [reply src/__tests__/index.js]: It was a dummy test created for coveralls purpose. We don't need it anymore. The reason for the deletion

**Review by zanio — COMMENTED** [review]
@OlawaleJoseph **Why is  6 test failing. After i pull your recent code from github that is what i get.**

image

**Comment by Dolapo47** [comment]
I cannot run seed on your pr, kindly look into it

**Comment by OlawaleJoseph** [comment]
> I cannot run seed on your pr, kindly look into it

No need to to run seed. Just run npm test

File: package.json [hunk 5]
```diff
@@ -4,8 +4,8 @@
...
-    "coverage": "nyc report --reporter=text",
+    "test": "npm run undo:test && npm run migrate:test && npm run seed:test && cross-env NODE_ENV=test nyc mocha --require @babel/register ./src/__tests__/*.js --timeout 20000",
```
Comment (Dolapo47) [inline package.json]: why this, we will have to conclude on the test script we will all use

**Review by Dolapo47 — DISMISSED** [review]
Good work

File: package.json (hunk 5 above)
Comment (OlawaleJoseph) [reply package.json]: This is to reset the database before running the .

**Review by zanio — COMMENTED** [review]
This is what i get when i tried to set persmission according to your `PR` 

image

Kindly run `npm run migrate:dev and npm run seed:dev` to seed the database

**Review by zanio — COMMENTED** [review]
Gud work. Your test are okay

File: package.json [hunk 6]
```diff
@@ -14,15 +14,16 @@
...
+    "undo:dev": "sequelize db:migrate:undo:all",
+    "undo:test": "cross-env NODE_ENV=test sequelize db:migrate:undo:all",
     "undo:prod": "cross-env NODE_ENV=prooduction sequelize db:migrate:undo",
     "seed:dev": "sequelize db:seed:all",
     "seed:test": "cross-env NODE_ENV=test sequelize db:seed:all"
   },
   "author": "Andela Simulations Programme",
   "license": "MIT",
   "dependencies": {
+    "@hapi/joi": "^15.1.1",
```
Comment (zanio) [inline package.json:26]: What is the use of this package. can you provide resource on the slack channel ?

File: src/index.js [hunk 7]
```diff
@@ -24,33 +21,19 @@ app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerDocument));
...
-    saveUninitialized: false,
-  }),
-);
-
 if (!isProduction) {
   app.use(errorHandler());
 }
 
-app.use(routes);
+app.use('/api/v1', routes);
```
Comment (zanio) [inline src/index.js]: Why was this remove from the routes folder and brought here

File: src/index.js [hunk 8]
```diff
@@ -59,33 +42,20 @@ app.use((req, res, next) => {
...
-    serverLog(err.stack);
-
     res.status(err.status || 500);
-
     res.json({
-      errors: {
-        message: err.message,
-        error: err,
-      },
+      success: false,
```
Comment (zanio) [inline src/index.js:50]: You are nt providing a status code in your response body.

File: src/validation/permissions.js [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+      attributes: [permission],
+      where: {
+        roleId,
+        resourceId
+      }
+    });
+    if (!foundPermission) {
+      throw new Error('You are not authorized to perform this operation');
+    }
+    // If Permission exists return the value of the permission
```
Comment (zanio) [inline src/validation/permissions.js]: Use js-doc instead

**Review by zanio — COMMENTED** [review]
Great job!. Implement those changes and get back to me for approval

File: package.json (hunk 6 above)
Comment (OlawaleJoseph) [reply package.json:26]: @hapi/joi is an npm package used for validation.

File: src/validation/permissions.js (hunk 9 above)
Comment (OlawaleJoseph) [reply src/validation/permissions.js]: The purpose of js-doc is for documentation. The above is just a comment to explain what is happening. I don't want to document if a variable exists or not

File: src/index.js (hunk 8 above)
Comment (OlawaleJoseph) [reply src/index.js:50]: During code review today, TTL asked that it should be changed. It is already in the response status

Comment (OlawaleJoseph) [reply src/index.js:50]: During code review today, TTL asked that it should be changed. It is already in the response status

File: src/index.js (hunk 7 above)
Comment (OlawaleJoseph) [reply src/index.js]: This is just due to preference. They both do the same thing

Comment (zanio) [reply src/index.js]: When it is merged with mine which resides in routes folder  won't that result to conflict

File: src/index.js (hunk 8 above)
Comment (zanio) [reply src/index.js:50]: You mean it should just be true or false

File: src/index.js (hunk 7 above)
Comment (OlawaleJoseph) [reply src/index.js]: No it won't. U can put yours in the folder and remove this from the index file

**Review by zanio — APPROVED** [review]
great work!

**Review by zanio — APPROVED** [review]
great work!

**Review by dewaleolaoye — APPROVED** [review]
Well Done!

**Review by ola357 — CHANGES_REQUESTED** [review]
welldone, @OlawaleJoseph 
1. please adhere to the database structure the team agreed upon.
2. use camel casing for the fields of the user model

File: .env.example [hunk 10]
```diff
@@ -1,9 +1,10 @@
-DB_USERNAME=postgres
-DB_PASSWORD=db_password
-DB_NAME=barefoot_nomad
-DB_HOST=127.0.0.1
-DB_TEST_NAME=barefoot_test
-PORT = 
+DB_USERNAME=database username
```
Comment (marusoft) [inline .env.example:1]: Why do you have duplicate DB_USERNAME? and so on

File: src/controllers/permissions.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+   * @returns {object } Updated Role information
+   * @memberof Permissions
+   * @type {object}
+   */
+  static async setPermissions(req, res, next) {
+    let updatedPermissions;
+    const { roleId } = req.params;
+    const { resourceId, ...permissions } = req.body;
+    try {
+      // Check if Resource Exists
```
Comment (marusoft) [inline src/controllers/permissions.js]: Remove comment

File: src/controllers/permissions.js [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+      await checkIfExistsInDb(Resource, resourceId, 'Resource does not exist');
+      // Check If Permission alreday exists in Db
```
Comment (marusoft) [inline src/controllers/permissions.js]: remove comment

File: src/index.js [hunk 13]
```diff
@@ -60,35 +47,20 @@ app.use((req, res, next) => {
...
+      message: err.message || 'Something went wrong',
     });
   });
 }
-
 // production error handler
```
Comment (marusoft) [inline src/index.js:55]: remove comment

File: src/index.js [hunk 14]
```diff
@@ -60,35 +47,20 @@ app.use((req, res, next) => {
...
-      message: err.message,
-      error: {},
-    },
-  });
-});
 
 // finally, let's start our server...
-const server = app.listen(process.env.PORT || 3000, () => {
+const server = app.listen(process.env.PORT, () => {
   serverLog(`Listening on port ${server.address().port}`);
```
Comment (marusoft) [inline src/index.js]: According to TTL, we are asked to change serverLog to console.log

File: src/middlewares/permissions.js [hunk 15]
```diff
@@ -0,0 +1,44 @@
+import Joi from '@hapi/joi';
+import { Resource } from '../models';
+import { permissionType, checkPermission, setPermissionSchema } from '../validation/permissions';
+
+export const permit = async (req, res, next) => {
+  const { roleId } = req.user;
+  const permission = permissionType(req);
+  try {
+    // Find resource by name in the db if it exists
```
Comment (marusoft) [inline src/middlewares/permissions.js]: remove comment

Comment (marusoft) [inline src/middlewares/permissions.js]: remove comment

File: src/validation/permissions.js [hunk 16]
```diff
@@ -0,0 +1,76 @@
+import Joi from '@hapi/joi';
+import { Role, Permission } from '../models';
+
+export const checkPermission = async (roleId, resourceId, permission) => {
+  try {
+    // Base case to check if there is no roleId given
+    if (!roleId) throw new Error('You are not authorized to perform this operation');
+    // Find Permission if exists
```
Comment (marusoft) [inline src/validation/permissions.js]: remove comment

File: src/validation/permissions.js [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+        resourceId
+      }
+    });
+    if (!foundPermission) {
+      throw new Error('You are not authorized to perform this operation');
+    }
+    // If Permission exists return the value of the permission
+    const isPermitted = foundPermission.dataValues[permission];
+    if (isPermitted) return isPermitted;
+    // If permission does not exist
```
Comment (marusoft) [inline src/validation/permissions.js]: remove comment

File: src/index.js (hunk 13 above)
Comment (marusoft) [inline src/index.js:55]: please remove comment

**Review by marusoft — CHANGES_REQUESTED** [review]
Awesome job @OlawaleJoseph  please implement these changes

File: src/controllers/permissions.js (hunk 11 above)
Comment (OlawaleJoseph) [reply src/controllers/permissions.js]: The comment is to explain the flow of the code.

File: src/index.js (hunk 14 above)
Comment (OlawaleJoseph) [reply src/index.js]: Ok

**Comment by marusoft** [comment]
cool

File: .env.example (hunk 10 above)
Comment (OlawaleJoseph) [reply .env.example:1]: This is a screenshot of my  `.env.example` file. There is no duplication 
Screenshot (34)

**Comment by marusoft** [comment]
ok

**Comment by ola357** [comment]
please have a sync with the team on why you have so many roles, why does a manager and requester have separate roles for example @OlawaleJoseph

**Comment by OlawaleJoseph** [comment]
> please have a sync with the team on why you have so many roles, why does a manager and requester have separate roles for example @OlawaleJoseph

That is the product spec

File: src/__tests__/dummyLogin.js [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+chai.use(chaiHttp);
+
+const { assert } = chai;
+
+describe('Login', () => {
+  it('Should generate Token', async () => {
+    const res = await chai
+      .request(server)
+      .post('/api/v1/auth/login')
+      .send({ email: 'abc123@gmail.com' });
```
Comment (Dolapo47) [inline src/__tests__/dummyLogin.js:23]: why should it generate token if no password was given?

Comment (OlawaleJoseph) [reply src/__tests__/dummyLogin.js:23]: This is a dummy login route for the purpose of generating token. This will be removed once the login functionality has been implemented

**Review by Dolapo47 — APPROVED** [review]
approved

**Review by marusoft — APPROVED** [review]
Awesome job @OlawaleJoseph.

File: src/routes/api/roles.js [hunk 19]
```diff
@@ -0,0 +1,12 @@
+import { Router } from 'express';
+import Users from '../../controllers/users';
+import validateSetRole from '../../middlewares/userRole';
+import { permit } from '../../middlewares/permissions';
+import isLoggedIn from '../../middlewares/login';
+
+const router = Router();
+
+router.patch('/users/:email/role', [isLoggedIn, validateSetRole, permit], Users.changeRole);
```
Comment (rovilay) [inline src/routes/api/roles.js]: this api should be in the user routes folder because its under `users`.

Comment (rovilay) [reply src/routes/api/roles.js]: passing email as a param is not a good practice instead use `userId`.

File: src/routes/api/users.js [hunk 20]
```diff
@@ -3,6 +3,7 @@ import { SendVerificationEmail, handleInvalidEmail, handleEmptyEmailBody } from
 import emailverification from '../../controllers/emailController';
 
 const router = Router();
+router.get('/', (req, res) => res.send('Users Route'));
```
Comment (rovilay) [inline src/routes/api/users.js]: What is the use of this route?

File: src/routes/index.js [hunk 21]
```diff
@@ -1,8 +1,28 @@
...
+router.use('/api/v1', routes);
+router.post('/api/v1/auth/login', async (req, res, next) => {
+  try {
+    const user = await findByEmail(req.body.email);
+    const { id, roleId } = user;
+    let token = await jwt.sign({ id, roleId }, process.env.JWT_SECRET);
+    token = `Bearer ${token}`;
+    res
+      .status(200)
+      .header('x-auth-access', token)
```
Comment (rovilay) [inline src/routes/index.js]: Please use `authorization` for the token instead of `x-auth-access`

File: src/models/Role.js [hunk 22]
```diff
@@ -0,0 +1,21 @@
+export default (sequelize, DataTypes) => {
+  const Role = sequelize.define('role', {
+    name: {
+      type: DataTypes.STRING,
+      allowNull: false
+    },
+    parentId: {
```
Comment (rovilay) [inline src/models/Role.js]: Please what is parentId?

File: src/middlewares/login.js [hunk 23]
```diff
@@ -0,0 +1,27 @@
+import jwt from 'jsonwebtoken';
+
+const isLoggedIn = async (req, res, next) => {
+  let token = req.headers['x-auth-access'];
+  if (!token) {
+    return res.status(401).json({
+      status: 'error',
+      message: 'You are not loggged in'
```
Comment (rovilay) [inline src/middlewares/login.js]: Please use a better error message like 'token required`

File: src/middlewares/login.js [hunk 24]
```diff
@@ -0,0 +1,27 @@
...
+    });
+  }
+  try {
+    token = token.slice(7);
```
Comment (rovilay) [inline src/middlewares/login.js]: I think you can split by space `' '` and use [1] to get the token as opposed to slice. Just a suggestion though.

File: src/middlewares/permissions.js [hunk 25]
```diff
@@ -0,0 +1,44 @@
+import Joi from '@hapi/joi';
+import { resource } from '../models';
+import { permissionType, checkPermission, setPermissionSchema } from '../validation/permissions';
+
+export const permit = async (req, res, next) => {
```
Comment (rovilay) [inline src/middlewares/permissions.js]: Can you explain the permisson concept and logic?

File: src/middlewares/permissions.js [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+  const { roleId } = req.user;
+  const permission = permissionType(req);
+  try {
+    // Find resource by name in the db if it exists
+    const foundResource = await resource.findOne({
+      where: {
+        name: req.url.split('/')[1]
```
Comment (rovilay) [inline src/middlewares/permissions.js]: why this? what are you trying to get from the `url`?

File: src/validation/permissions.js [hunk 27]
```diff
@@ -0,0 +1,76 @@
+import Joi from '@hapi/joi';
```
Comment (rovilay) [inline src/validation/permissions.js]: Please move the validation folder into the `helpers` folder. It makes the reference easier and file structure better.

File: src/controllers/permissions.js [hunk 28]
```diff
@@ -0,0 +1,64 @@
+import { checkIfExistsInDb } from '../utils/searchDb';
```
Comment (rovilay) [inline src/controllers/permissions.js]: please follow the `emailController` filename structure.

File: src/controllers/permissions.js [hunk 29]
```diff
@@ -0,0 +1,64 @@
...
+   * @param {object} res object
+   * @param {method} next method
+   * @returns {object } Updated Role information
+   * @memberof Permissions
+   * @type {object}
+   */
+  static async setPermissions(req, res, next) {
+    let updatedPermissions;
+    const { roleId } = req.params;
+    const { resourceId, ...permissions } = req.body;
```
Comment (rovilay) [inline src/controllers/permissions.js]: why are you spreading permissions here?

File: src/controllers/users.js [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+ * @param {method} next method
+ * @returns {object } Sets Role for a given user
+ * @memberof Roles
+ * @type {object}
+ */
+  static async changeRole(req, res, next) {
+    const { email, roleId } = req.body;
+    try {
+      await findByEmail(email);
+      await checkIfExistsInDb(role, roleId, 'Role does not exist');
```
Comment (rovilay) [inline src/controllers/users.js]: checking if role exists is better done in a middleware. Your controller should only contain logic to actually update role in the db.

File: src/middlewares/permissions.js (hunk 25 above)
Comment (OlawaleJoseph) [reply src/middlewares/permissions.js]: Yes I can. I have created a google doc for the team on it already.

File: src/controllers/permissions.js (hunk 29 above)
Comment (OlawaleJoseph) [reply src/controllers/permissions.js]: The Super Admin can set more than one permission for a role on a resource. The input is
{
  resourceId: 3,
  add: true,
  read: true
}
So I am using the spread operators to get all possible permission input

File: src/validation/permissions.js (hunk 27 above)
Comment (OlawaleJoseph) [reply src/validation/permissions.js]: Done

File: src/middlewares/login.js (hunk 24 above)
Comment (OlawaleJoseph) [reply src/middlewares/login.js]: Done. Thanks

File: src/middlewares/login.js (hunk 23 above)
Comment (OlawaleJoseph) [reply src/middlewares/login.js]: Done

File: src/models/Role.js (hunk 22 above)
Comment (OlawaleJoseph) [reply src/models/Role.js]: Roles can inherit permissions from other roles( Hierarchical role-based access control), parentId referenced the parent role. 
I will be removing this functionality as I have come to the conclusion that it isn't important

File: src/routes/index.js (hunk 21 above)
Comment (OlawaleJoseph) [reply src/routes/index.js]: Done

File: src/routes/api/users.js (hunk 20 above)
Comment (OlawaleJoseph) [reply src/routes/api/users.js]: A mistake i overlooked. Removed.

File: src/routes/api/roles.js (hunk 19 above)
Comment (OlawaleJoseph) [reply src/routes/api/roles.js]: Ok. Adjusted

File: src/controllers/users.js (hunk 30 above)
Comment (OlawaleJoseph) [reply src/controllers/users.js]: Ok. Thanks for the insight

**Comment by rovilay** [comment]
@OlawaleJoseph good work so far. 
I noticed that if I put a space in the email i.e 
`{
	"roleId": 2,
	"email": " abc123@gmail.com"
}`

it returns this error ` "User with the given email does not exist"`
you should trim the email for spaces.

**Comment by rovilay** [comment]
@OlawaleJoseph if I don't provide an email in the `req.body` I get the error message `Invalid email`, the error should be `email is required`

**Comment by rovilay** [comment]
Also, your route name should be `/users/roles` not `/users/user/roles`. since we've decided not to use `:userId`.

**Comment by rovilay** [comment]
If I'm trying to upgrade a user to a role that the user is currently on, I should get a message that the user is already in that role i.e if I am trying to upgrade a `travel admin` to a `travel admin` I should receive the appropriate message.

File: src/helpers/DevNote [hunk 31]
```diff
@@ -0,0 +1,6 @@
+To protect a route, kindly import the permit middleware from the users file in the middlewares folder.
```
Comment (rovilay) [inline src/helpers/DevNote]: if you want to put a note, you can either put it in the `ReadMe.md` file or make this `DevNote` file a `DevNote.md` file.

File: src/helpers/default.js [hunk 32]
```diff
@@ -1,3 +1,11 @@
 /* eslint-disable import/prefer-default-export */
 // eslint-disable-next-line no-useless-escape
 export const regexForEmail = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+
+export const roleIds = {
+  'Super Administrator': 1,
```
Comment (rovilay) [inline src/helpers/default.js]: I remember mentioning that you should change this to `camelCase` without the string.

Comment (OlawaleJoseph) [reply src/helpers/default.js]: My apologies for this. It has been fixed

File: src/migrations/20190826132443-create-Role.js [hunk 33]
```diff
@@ -0,0 +1,23 @@
+module.exports = {
```
Comment (rovilay) [inline src/migrations/20190826132443-create-Role.js]: why did you delete this migraiton file?

File: src/routes/api/index.js [hunk 34]
```diff
@@ -3,6 +3,6 @@ import users from './users';
```
Comment (rovilay) [inline src/routes/api/index.js:3]: Why did you delete the `models/users.js` file?

File: src/seeders/roles.js [hunk 35]
```diff
@@ -0,0 +1,36 @@
+module.exports = {
```
Comment (rovilay) [inline src/seeders/roles.js]: Why did you delete the `create-users` seeders

**Comment by rovilay** [comment]
@OlawaleJoseph good job so far. Although, I mentioned to you that a `super admin` should not be able to change his/her role. It should require the help of another super admin to do that.

File: src/seeders/roles.js (hunk 35 above)
Comment (OlawaleJoseph) [reply src/seeders/roles.js]: I only renamed it, when sequelize required order.

File: src/migrations/20190826132443-create-Role.js (hunk 33 above)
Comment (OlawaleJoseph) [reply src/migrations/20190826132443-create-Role.js]: I only renamed it.

File: src/seeders/roles.js (hunk 35 above)
Comment (rovilay) [reply src/seeders/roles.js]: ok

File: src/routes/api/index.js (hunk 34 above)
Comment (OlawaleJoseph) [reply src/routes/api/index.js:3]: I renamed only renamed it to avoid  sequelize errror

**Comment by OlawaleJoseph** [comment]
> @OlawaleJoseph good job so far. Although, I mentioned to you that a `super admin` should not be able to change his/her role. It should require the help of another super admin to do that.

I am quite positive I implemented this.

File: src/migrations/20190826132443-create-Role.js (hunk 33 above)
Comment (rovilay) [reply src/migrations/20190826132443-create-Role.js]: Why will you rename a migration file? It can result in issues when running migrations for other developers.

File: src/routes/api/index.js (hunk 34 above)
Comment (rovilay) [reply src/routes/api/index.js:3]: how does it cause sequelize error?

Comment (OlawaleJoseph) [reply src/routes/api/index.js:3]: The order of arrangement in the migrations folder. Sequelize was throwing error because it was trying to create role roles which has a foreign key from users

File: src/migrations/20190826132443-create-Role.js (hunk 33 above)
Comment (OlawaleJoseph) [reply src/migrations/20190826132443-create-Role.js]: I never knew this

File: src/routes/api/index.js (hunk 34 above)
Comment (rovilay) [reply src/routes/api/index.js:3]: if they were properly done you shouldn't have that issue.

**Comment by OlawaleJoseph** [comment]
> > @OlawaleJoseph good job so far. Although, I mentioned to you that a `super admin` should not be able to change his/her role. It should require the help of another super admin to do that.
> 
> I am quite positive I implemented this.

**Comment by OlawaleJoseph** [comment]
> @OlawaleJoseph good job so far. Although, I mentioned to you that a `super admin` should not be able to change his/her role. It should require the help of another super admin to do that.

THis has been implemented. Below is a snapshot of the error admin who tries to change his/her role will receive.
Screenshot (45)

**Comment by rovilay** [comment]
@OlawaleJoseph this is what I'm getting
Screenshot 2019-09-06 at 10 04 18 AM

**Comment by OlawaleJoseph** [comment]
> @OlawaleJoseph this is what I'm getting
> Screenshot 2019-09-06 at 10 04 18 AM

I think the reason for this is because you are using an old token, I passed in an additional parameter into the token which is not present in the token you are using to access, this route. Kindly generate a new token

**Comment by rovilay** [comment]
Please fix conflicts.

**Comment by rovilay** [comment]
> > @OlawaleJoseph this is what I'm getting
> > Screenshot 2019-09-06 at 10 04 18 AM
> 
> I think the reason for this is because you are using an old token, I passed in an additional parameter into the token which is not present in the token you are using to access, this route. Kindly generate a new token

I generated a new token
Screenshot 2019-09-06 at 11 16 54 AM

File: src/__tests__/services/mail.js [hunk 36]
```diff
@@ -16,69 +16,68 @@ chai.use(chaiHttp);
...
-            try {
-                sgMail.setApiKey(fakeApiKey);
-            } catch (err) {
-                throw err;
-            }
-        });
+  describe('Should fail to send mail', async () => {
+    beforeEach(async () => {
+      try {
+        sgMail.setApiKey(fakeApiKey);
```
Comment (rovilay) [inline src/__tests__/services/mail.js:22]: why are you editing the indentations?

File: src/config/config.js [hunk 37]
```diff
@@ -6,8 +6,7 @@ module.exports = {
     password: process.env.DB_PASSWORD,
     database: process.env.DB_NAME,
     host: process.env.DB_HOST,
-    dialect: 'postgres',
-    operatorsAliases: false
```
Comment (rovilay) [inline src/config/config.js:10]: why did you remove `operatorsAliases`?

File: src/__tests__/services/mail.js (hunk 36 above)
Comment (OlawaleJoseph) [reply src/__tests__/services/mail.js:22]: I linted before pushing. I guess eslint did that as they were many eslint errors

File: src/config/config.js (hunk 37 above)
Comment (OlawaleJoseph) [reply src/config/config.js:10]: operatorsAliases has been deprecated and shows as a warning in the console.

**Comment by rovilay** [comment]
@OlawaleJoseph the team have to stick to one type of indentation and it should be in the eslint config.

Yes. That is what I used. The eslint config for the team

**Comment by rovilay** [comment]
but I can't find any indentation rule in your eslint config

**Comment by OlawaleJoseph** [comment]
> @OlawaleJoseph the team have to stick to one type of indentation and it should be in the eslint config.
> 
Yes. That is what I used. The eslint config for the team

File: src/config/config.js (hunk 37 above)
Comment (rovilay) [reply src/config/config.js:10]: ok

**Comment by rovilay** [comment]
Please add it.

**Comment by OlawaleJoseph** [comment]
> but I can't find any indentation rule in your eslint config

All I did was lint. I have no different configuration. The other thing that comes to mind is prettier extension in my code editor.
